apps/DesktopStreamer/DesktopWindowsModel.mm

```diff
@@ -0,0 +1,336 @@
+    CGRect rect;
+    CGRectMakeWithDictionaryRepresentation( bounds, &rect );
+    CFRelease( windowList );
+
+    return QRect( CGRectGetMinX( rect ), CGRectGetMinY( rect ),
+                  CGRectGetWidth( rect ), CGRectGetHeight( rect ));
+}
+
+} // namespace
+
+
+@interface AppObserver : NSObject
+{
+    DesktopWindowsModel* parent;
+}
+- (void)setParent:(DesktopWindowsModel*)parent_;
+@end
+
+@implementation AppObserver
+
+- (id) init
+{
+    self = [super init];
+
+    NSWorkspace *workspace = [NSWorkspace sharedWorkspace];
+    NSNotificationCenter* center = [workspace notificationCenter];
+    [center addObserver:self
+            selector:@selector(newApplication:)
+            name:NSWorkspaceDidLaunchApplicationNotification
+            object:Nil];
+    [center addObserver:self
+            selector:@selector(closedApplication:)
+            name:NSWorkspaceDidTerminateApplicationNotification
+            object:Nil];
+
+    return self;
+}
+
+- (void) dealloc
+{
+    NSWorkspace *workspace = [NSWorkspace sharedWorkspace];
+    NSNotificationCenter* center = [workspace notificationCenter];
+    [center removeObserver:self
+            name:NSWorkspaceDidLaunchApplicationNotification
+            object:Nil];
+    [center removeObserver:self
+            name:NSWorkspaceDidTerminateApplicationNotification
+            object:Nil];
+
+    [super dealloc];
+}
+
+- (void)setParent:(DesktopWindowsModel*)parent_
+{
+    parent = parent_;
+}
+
+- (void)newApplication:(NSNotification*) notification
+{
+    NSRunningApplication* app =
+            [[notification userInfo] objectForKey:@"NSWorkspaceApplicationKey"];
+    parent->addApplication( app );
+}
+
+- (void)closedApplication:(NSNotification*) notification
+{
+    NSRunningApplication* app =
+            [[notification userInfo] objectForKey:@"NSWorkspaceApplicationKey"];
+    parent->removeApplication( app );
+}
+
+@end
+
+
+class DesktopWindowsModel::Impl
+{
+public:
+    Impl( DesktopWindowsModel& parent )
+        : _parent( parent )
+        , _observer( [[AppObserver alloc] init] )
+    {
+        [_observer setParent:&parent];
+        reloadData();
+    }
+
+    ~Impl()
+    {
+        [_observer release];
+    }
+
+    enum TupleValues
+    {
+        APPNAME,
+        WINDOWID,
+        WINDOWIMAGE
+    };
+
+    void addApplication( NSRunningApplication* app )
+    {
+        _parent.beginInsertRows( QModelIndex(), _data.size(), _data.size( ));
+        for( ;; )
+        {
+            CFArrayRef windowList =
+                CGWindowListCopyWindowInfo( kCGWindowListOptionOnScreenOnly|
+                                            kCGWindowListExcludeDesktopElements,
+                                            kCGNullWindowID );
+            if( _addApplication( app, windowList ))
+                break;
+            // notification is not sync'd with CGWindowListCopyWindowInfo
+            sleep( 1 );
```

Comment:
Hard to say. While you're going back to DeskopStreamer after opening a new application, the sleep is over :)